packages/apollo-federation/src/composition/validate/sdl/uniqueFieldDefinitionNames.ts

```diff
@@ -0,0 +1,180 @@
+import {
+  ASTVisitor,
+  NameNode,
+  GraphQLError,
+  InputObjectTypeDefinitionNode,
+  InterfaceTypeDefinitionNode,
+  ObjectTypeDefinitionNode,
+  InputObjectTypeExtensionNode,
+  InterfaceTypeExtensionNode,
+  ObjectTypeExtensionNode,
+  GraphQLNamedType,
+  isObjectType,
+  isInterfaceType,
+  isInputObjectType,
+} from 'graphql';
+import { SDLValidationContext } from 'graphql/validation/ValidationContext';
+import { TypeMap } from 'graphql/type/schema';
+import Maybe from 'graphql/tsutils/Maybe';
+import { diffTypeNodes } from '../../utils';
+
+type NodeTypesRequiringUniqueFields =
+  | TypeDefinitionsRequiringUniqueFields
+  | TypeExtensionsRequiringUniqueFields;
+
+type TypeDefinitionsRequiringUniqueFields =
+  | InputObjectTypeDefinitionNode
+  | InterfaceTypeDefinitionNode
+  | ObjectTypeDefinitionNode;
+
+type TypeExtensionsRequiringUniqueFields =
+  | InputObjectTypeExtensionNode
+  | InterfaceTypeExtensionNode
+  | ObjectTypeExtensionNode;
+
+export function duplicateFieldDefinitionNameMessage(
+  typeName: string,
+  fieldName: string,
+): string {
+  return `Field "${typeName}.${fieldName}" can only be defined once.`;
+}
+
+export function existedFieldDefinitionNameMessage(
+  typeName: string,
+  fieldName: string,
+): string {
+  return `Field "${typeName}.${fieldName}" already exists in the schema. It cannot also be defined in this type extension.`;
+}
+
+/**
+ * Unique field definition names
+ *
+ * A GraphQL complex type is only valid if all its fields are uniquely named.
+ * Modified to permit duplicate field names on value types.
+ */
+export function UniqueFieldDefinitionNames(
+  context: SDLValidationContext,
+): ASTVisitor {
+  const schema = context.getSchema();
+  const existingTypeMap: TypeMap = schema
+    ? schema.getTypeMap()
+    : Object.create(null);
+  interface FieldToNameNodeMap {
+    [fieldName: string]: NameNode;
+  }
+  const knownFieldNames: {
+    [typeName: string]: FieldToNameNodeMap;
+  } = Object.create(null);
+
+  const possibleValueTypes: {
+    [key: string]: NodeTypesRequiringUniqueFields | undefined;
+  } = Object.create(null);
+
+  // Maintain original functionality for type extensions, but substitute our
+  // more permissive validator for base types to allow value types
+  return {
+    InputObjectTypeExtension: checkFieldUniqueness,
+    InterfaceTypeExtension: checkFieldUniqueness,
+    ObjectTypeExtension: checkFieldUniqueness,
+    InputObjectTypeDefinition: checkFieldUniquenessExcludingValueTypes,
+    InterfaceTypeDefinition: checkFieldUniquenessExcludingValueTypes,
+    ObjectTypeDefinition: checkFieldUniquenessExcludingValueTypes,
+  };
+
+  function checkFieldUniqueness(node: TypeExtensionsRequiringUniqueFields) {
+    const typeName = node.name.value;
+
+    if (!knownFieldNames[typeName]) {
+      knownFieldNames[typeName] = Object.create(null);
+    }
+
+    if (node.fields) {
```

Comment:
What if this was:

```suggestion
    if (!node.fields) {
      return false;
    }
```

... with the remainder of this block shifted two spaces to the left (less nested!).

Reply:
👍 
Resolved in 7d522da7975f640ef577b19be6cab88868119da5